### PR TITLE
185337946 multiple client mci ids

### PR DESCRIPTION
**Pull request by ttoomey**
reference:
https://www.pivotaltracker.com/story/show/185337946

* adjust db constraint to allow clients to have more than one MCI ID
* made use of the existing `client.mci_id` method where we need one MCI ID
* Adjust MergeClientsJob to merge MCI IDs
* Adjust ClientProcessor to check all MCI IDs

### CODE REVIEW

File: db/warehouse/migrate/20230613190449_adjust_external_ids_uniq_idx.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class AdjustExternalIdsUniqIdx < ActiveRecord::Migration[6.1]
+  disable_ddl_transaction!
+  def change
+    # allow more than one client MCI ID
+    add_index :external_ids,
+        [:source_id, :source_type, :remote_credential_id],
+        unique: true,
+        name: 'uidx_external_ids_source_value',
+        where: "(namespace <> 'ac_hmis_mci')",
+        algorithm: :concurrently
```
Comment (gigxz) [inline db/warehouse/migrate/20230613190449_adjust_external_ids_uniq_idx.rb]: So this is saying, every namespace EXCEPT mci ids must have no more than 1 value per remote credential per client? If we run into another ExternalID that allows multiple values, how would we handle that? I don't know how likely that is, so maybe this is fine for now.

File: db/warehouse/migrate/20230613190449_adjust_external_ids_uniq_idx.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class AdjustExternalIdsUniqIdx < ActiveRecord::Migration[6.1]
+  disable_ddl_transaction!
```
Comment (gigxz) [inline db/warehouse/migrate/20230613190449_adjust_external_ids_uniq_idx.rb:2]: I've had issues deploying migrations that do this, @eanders it looked like you encountered it on another staging deployment. How should we handle it?

File: drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/mci.rb [hunk 3]
```diff
@@ -106,7 +106,7 @@ def create_mci_id(client)
     def update_client(client)
       raise(Error, 'Client needs to be saved first') unless client.persisted?
 
-      external_id = get_external_id(client)
+      external_id = client.mci_id
```
Comment (gigxz) [inline drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/mci.rb]: It looks like `get_external_id` returned an ExternalId, but `mci_id` returns a string

**Review by gigxz — COMMENTED** [review]
This looks good, tested all the MCI features and it's working fine. The update_client request isn't being used (yet? ever? tbd) but it looks like there is a bug there that should be fixed before merging.

File: db/warehouse/migrate/20230613190449_adjust_external_ids_uniq_idx.rb (hunk 1 above)
Comment (ttoomey) [reply db/warehouse/migrate/20230613190449_adjust_external_ids_uniq_idx.rb]: > So this is saying, every namespace EXCEPT mci ids must have no more than 1 value per remote credential per client? If we run into another ExternalID that allows multiple values, how would we handle that? I don't know how likely that is, so maybe this is fine for now.

Yes, that's right. If we need another namespace that is non-unique, we could expand this condition or perhaps we add a dedicated flag. 

Looking at it again I realize that namespace can be null which doesn't match the condition 🤦 - I'll adjust that

File: drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/mci.rb (hunk 3 above)
Comment (ttoomey) [reply drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/mci.rb]: Thats a great spot, thank you!

**Review by gigxz — APPROVED** [review]

File: db/warehouse/migrate/20230613190449_adjust_external_ids_uniq_idx.rb (hunk 2 above)
Comment (gigxz) [reply db/warehouse/migrate/20230613190449_adjust_external_ids_uniq_idx.rb:2]: I guess we'll just try it again, we know how to work around it if it times out again

Comment (eanders) [reply db/warehouse/migrate/20230613190449_adjust_external_ids_uniq_idx.rb:2]: Yeah, sorry, this is also on my list to look at.  It _should_ be totally fine to disable the transaction, it jus means it won't lock the table for reading/writing.

**Comment by ttoomey** [comment]
It seems like CI isn't working, holding off on the merge

**Comment by eanders** [comment]
@ttoomey I _think_ I got CI working [here](https://github.com/greenriver/hmis-warehouse/pull/3169), still not sure what the issue is, but the changes I made seem to have helped.  If that passes after the merge to `release-78`, I'll merge into `release-79` and we should be back to normal.

**Comment by eanders** [comment]
@ttoomey it does appear tests are working now, just one broken one.

**Comment by gigxz** [comment]
Thanks @eanders . One of the test failures is fixed by https://github.com/greenriver/hmis-warehouse/pull/3175.

**Comment by ttoomey** [comment]
> @ttoomey it does appear tests are working now, just one broken one.

Thanks @gigxz and @eanders 

The second test failure was strange. While I could reproduce it locally, it wan't reliable; the failure seemed random. Looking at the code in question ([`MergeClientJob.dedup`](https://github.com/greenriver/hmis-warehouse/blob/release-79/drivers/hmis/app/jobs/hmis/merge_clients_job.rb#L115)), I couldn't see the bug but I did notice the flow was more complicated than seemed necessary. I refactored the method and the tests pass reliably for me now.

[My refactor](https://github.com/greenriver/hmis-warehouse/blob/0ce0e88c576381be86e4a46419e2d05341c25d75/drivers/hmis/app/jobs/hmis/merge_clients_job.rb#L117) could definitely use a review. I've also changed usage of `destroy` to use the bang-variant to catch failures

File: drivers/hmis/app/jobs/hmis/merge_clients_job.rb [hunk 4]
```diff
@@ -109,29 +110,24 @@ def merge_custom_data_elements
...
       end
     end
 
-    def dedup(relation, keepers: [])
-      pairs = client_to_retain.send(relation).to_a.combination(2)
-
-      pairs.each do |pair|
-        equivalent = if pair[0].respond_to?(:equal_for_merge?)
-          pair[0].equal_for_merge?(pair[1])
+    def dedup(records, keepers: [])
```
Comment (gigxz) [inline drivers/hmis/app/jobs/hmis/merge_clients_job.rb:117]: @ttoomey this looks correct to me. thank you for doing all this to fix the tests

**Review by gigxz — APPROVED** [review]

**Pull request merged**